cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,685 @@
+                  // For all row groups
+                  std::for_each(rg_indices.cbegin(), rg_indices.cend(), [&](auto const rg_index) {
+                    // For all column chunks
+                    std::for_each(
+                      column_schemas.begin(), column_schemas.end(), [&](auto const schema_idx) {
+                        auto& col_meta = get_column_metadata(rg_index, src_index, schema_idx);
+
+                        // Get bloom filter offsets and sizes
+                        bloom_filter_offsets[chunk_count] = col_meta.bloom_filter_offset;
+                        bloom_filter_sizes[chunk_count]   = col_meta.bloom_filter_length;
+
+                        // Map each column chunk to its source index
+                        chunk_source_map[chunk_count] = src_index;
+                        chunk_count++;
+                      });
+                  });
+                });
+
+  // Do we have any bloom filters
+  if (std::any_of(bloom_filter_offsets.cbegin(),
+                  bloom_filter_offsets.cend(),
+                  [](auto const offset) { return offset.has_value(); })) {
+    // Vector to hold bloom filter data
+    std::vector<rmm::device_buffer> bloom_filter_data(num_chunks);
+
+    // Wait on bloom filter data read tasks
+    read_bloom_filter_data_async(sources,
+                                 num_chunks,
+                                 bloom_filter_data,
+                                 bloom_filter_offsets,
+                                 bloom_filter_sizes,
+                                 chunk_source_map,
+                                 stream)
+      .wait();
+
+    // Return bloom filter data
+    return bloom_filter_data;
+  }
+
+  // Return empty vector
+  return {};
+}
+
+std::vector<Type> aggregate_reader_metadata::get_parquet_types(
+  host_span<std::vector<size_type> const> row_group_indices,
+  host_span<int const> column_schemas) const
+{
+  std::vector<Type> parquet_types(column_schemas.size());
+  // Find a source with at least one row group
+  auto const src_iter = std::find_if(row_group_indices.begin(),
+                                     row_group_indices.end(),
+                                     [](auto const& rg) { return rg.size() > 0; });
+  CUDF_EXPECTS(src_iter != row_group_indices.end(), "");
+
+  // Source index
+  auto const src_index = std::distance(row_group_indices.begin(), src_iter);
+  std::transform(column_schemas.begin(),
+                 column_schemas.end(),
+                 parquet_types.begin(),
+                 [&](auto const schema_idx) {
+                   // Use the first row group in this source
+                   auto constexpr row_group_index = 0;
+                   return get_column_metadata(row_group_index, src_index, schema_idx).type;
+                 });
+
+  return parquet_types;
+}
+
+std::optional<std::vector<std::vector<size_type>>> aggregate_reader_metadata::apply_bloom_filters(
```

Comment:
do we need the optional here, i.e. is nullopt any different from returning an empty vector?

Reply:
We could but this makes it easier to return the output optional as-is in `filter_row_groups()` where this function is called from.